### ISSUE TITLE
Extend MA0002 to warn for any comparer

### ISSUE THREAD

**Issue by LPeter1997** [meziantou/Meziantou.Analyzer#314]
It would be very useful for library authors to have a generalized version of MA0002, that would warn on anything missing the `IComparer<T>` or `IEqualityComparer<T>`. In a generic context, writing a custom collection, it is easy to forget passing in the appropriate comparer when calling `SequenceEquals(...)` or constructing a `HashSet<T>`. It would be neat, if I could be warned, when I have the possibility to use an overload with a comparer, but I don't.

**Comment by meziantou** [comment]
Currently, MA0002 check for `IComparer<T>` and `IEqualityComparer<T>` when `T` is `string`. If I well understand, you want to extend (a new rule or using a config) the rule to report for any `T`?

Note that this may be very verbose, and the fix will often be `EqualityComparer<T>.Default` or `Comparer<T>.Default`

**Comment by LPeter1997** [comment]
Exactly, I think it would be useful to check for any `T` and not just `string`s. And yes, this could get very verbose, but right now I'm authoring libraries where I can't miss a single case. For nullable cases (like `IComparer<T>?`) , explicitly passing in `null` could also suppress the warning. I believe a separate rule would be beneficial here, since many would want this off, while keeping the string one on.

**Comment by meziantou** [comment]
It should be a different rule, disabled by default because it is specific and verbose.

Do you think the rule should consider any overload with an additional `IComparer<T>` parameter whatever the `T`? Or should it use the same rule as MA0002?

case 1:

````c#
A(new object()); // should it report?

void A(object o) { }
void A(object o, IComparer<int> comparer) { }
````

case 2
    
````c#
A(null); // should it report?

void A(IComparer<int> comparer1) { }
void A(IComparer<int> comparer1, IComparer<int> comparer2) { }
````

case 3
    
````c#
var set = new HashSet<int>(); // report?
set.Contains(0); // report? (system.linq.enumerable.contains(this IEnumerable<T>, T, IEqualityComparer<T>)
````

**Comment by LPeter1997** [comment]
I believe that case 1 should report, case 2 shouldn't and case 3 should only report for the constructor. My rationale being:
 * I might want to expose convenience methods to the user, where it's their responsibility to pass in the comparer, so case 1 should be a warning for me.
 * In case 2 I've explicitly stated my comparer is null. Even if there is another overload, I couldn't expect the analyzer to somehow quantify which is the "better" case here and I don't think more comparers should count as "better" overloads here.
 * For case 3, I've missed the constructor with the comparer, so it should report. I believe dealing with the next line would be the trickiest in the analyzer, but I think it shouldn't report, because I refer to the member function, not the LINQ extension function. If I was referring to the extension function, it should consider the comparer overloads (like calling `.SequenceEqual(...)` there).

**Comment by meziantou** [comment]
I'm fine with this behavior. Would you like to submit a PR for this new analyzer?

**Comment by LPeter1997** [comment]
Definitely, I'll poke around the source in a few days.
